### ISSUE TITLE
Full screen mode

### ISSUE THREAD

**Issue by franzlst** [DLR-RM/RAFCON#832]
For presentation purpose or optimal screen usage a full screen mode would be useful.

Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-01-20 06:32:47+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Is this needed, considering the upcoming feature of undockable sidebars?


Originally created by @franzlst  (Franz.Steinmetz@dlr.de) at 2016-02-19 12:35:23+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
We will still have the menu bar visible all the time. I would still do this but it has low priority.


Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-02-19 14:20:14+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Solution in commit 551f35c.
Please close this request if everything is fine. 


Originally created by Benno  at 2016-09-05 15:13:56+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
So I can confirm the undock feature and especially the redock is broken by commit 5fdcec07. Means, anytime I undock a window with sled 64Bit or 32Bit machine 
- [x] the undocked window is poping up in the background
- [x] the undocked window can not be expanded to full screen by its respective button 
- [x] and also can not be redocked by the button next to it.

The poping up in the background is not very intuitive and the not usable buttons are something else, a bug.

Maybe we implement an unit test to be annoyed by this once. -> But that is an issue that concerns the  undocking feature it self.


Originally created by @Rbelder  at 2016-09-05 16:49:50+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@Rbelder: Could you please move your comment to the appropriate issue?

@Benno: This is definitely a cool feature. Here some things that need to be tackled:
- [x] Make it possible to get out of fullscreen mode with the escape key. In my opinion, this should be hardcoded, thus not configurable in the GUI config. The problem is: You can go to fullscreen mode using the menu, but if you haven't configured the shortcut correctly, you cannot go back.
- [x] Fullscreen mode does not respect the window in which RAFCON is open, but always uses the first window
- [x] On my OpenSUSE Leap Gnome3 system, when going out of fullscreen mode, the window is minimized
- [x] When you go out of fullscreen mode for the 2nd time, to [+ STATE] buttons move above the graphical editor and hide the graphical editor tabs


Originally created by @franzlst  (Franz.Steinmetz@dlr.de) at 2016-09-06 06:07:42+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Refactored in commit  a0b5b92.
I hope that it fixes the problems 1,3 and 4 in comment above...
btw I didnt had the problem number 3 and I only have one monitor at the moment so I wont be able to verify/ fix number 2 but I will as soon as possible ;) 

@Rbelder: As I said yesterday, I cannot maximize them too, but my undocked windows appears in the foreground :)   


Originally created by Benno  at 2016-09-06 07:52:27+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@franzlst maybe you tell me fully what you mean with appropriate?!
When I read the commit message of commit 5fdcec07 there is the only choice that this issue is causing the problem, especially the new properties of the undocked windows. If you like to open an issue that collects all bugs this issue here is causing you can do it but that would not be appropriate from my perspective.
I also updated the comment above with check boxes and pointed out that the last sentence concerns to maybe open a separate issue for a unit test that covers caused problems related to this feature implementation.

@Benno Please take a deeper lock into commit 5fdcec07. In my case all reported problems starting with this commit.


Originally created by @Rbelder  at 2016-09-06 08:08:48+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@Rbelder If your issues are related with that commit okay (I don't think so), otherwise I think you are referring to #136. Remember: fullscreen !=  maximize (this issue is about fullscreen mode for the graphical editor and not at all related to sidebars, neither their maximization) 

@Benno I have a spare monitor in my office which I can borrow you for some time


Originally created by @franzlst  (Franz.Steinmetz@dlr.de) at 2016-09-06 08:16:36+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@franzlst Please give me some thrust. It is maybe hard to believe but I checkout the commit before f7bf8c7c and the commit 5fdcec0 ... and it starts with this one. And was doing this before starting to write any comment. Thanks in advance.


Originally created by @Rbelder  at 2016-09-06 08:27:03+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@Rbelder Good possible that your problems starts with commit  5fdcec0 since i changed the undocked_window.glade file in it...but unfortunately it is hard for me to fix this because everything seems to be fine on my system except the maximizing. However, I will take a deeper look into it.

Added signal handler and fullscreen verification in commit 0b6dc18...


Originally created by Benno  at 2016-09-06 08:47:26+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Problem number 2 solved by commit 9a861cb.


Originally created by Benno  at 2016-09-06 15:07:37+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
So I can confirm the problems on sled 32- and 64-bits are solved. Good work.


Originally created by @Rbelder  at 2016-09-07 09:06:18+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Refactored in commit d7dfa66, since it was not cleanly working in KDE ;)
Please advise if there still any problem or new ones appeared.


Originally created by Benno  at 2016-09-07 11:55:48+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Sorry, I found one more thing. 

Lets say a question, is it the aimed behavior that the undocked windows stay in the foreground and the main window always behind (background) the undocked windows? Other occurrence, if I overlap (not even fully) one of the undocked windows with e.g. my pycharm window all the undocked windows disappear till I focus the main window which is again behind the undocked windows which also can not be changed with second or third focus (click) of the main window.

So, at the moment, I can edit the state machine with all 4 windows separated. If I have a undocked window overlapping the main window/graphical editor I first have to move it somewhere else to use the full area of the graphical editor.

Added later: I was writing about 32-bit and 64-bit sled11 systems with kde4 desktop.


Originally created by @Rbelder  at 2016-09-08 09:32:51+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
> Lets say a question, is it the aimed behavior that the undocked windows stay in the foreground and the main window always behind (background) the undocked windows? 

Yes! Otherwise it wouldn't be possible to bring the sidebars to the front, if they are hidden by the main window, as they have no entries in the task bar. (Well except moving the main window, which is tedious).

> Other occurrence, if I overlap (not even fully) one of the undocked windows with e.g. my pycharm window all the undocked windows disappear till I focus the main window which is again behind the undocked windows which also can not be changed with second or third focus (click) of the main window.

I don't get what you are saying. So if a sidebar is half over PyCharm and PyCharm is being focused, it will hide the sidebar (which is totally fine). If you then click on the main window _or_ the sidebar, _all_ RAFCON windows/sidebars will be arranged to the front again. Is this what you mean? If so, then this is also perfectly fine in my opinion.

> So, at the moment, I can edit the state machine with all 4 windows separated. If I have a undocked window overlapping the main window/graphical editor I first have to move it somewhere else to use the full area of the graphical editor.

This is the same situation as in your first question, right?


Originally created by @franzlst  (Franz.Steinmetz@dlr.de) at 2016-09-08 09:41:34+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Summary of the sidebar situation:
Using gtk.gdk.WINDOW_TYPE_HINT_UTILITY: 
- undocked window cannot be maximized in GNOME anymore
- undocked window disappears if main window loses focus in KDE (reappears if main window receives focus again)

Rico suggested to introduce some "taskbars" as like in Pycharm to provide the separate restore functionallity.   


Originally created by Benno  at 2016-09-08 11:31:23+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Frankly speaking, I think it won't get any better than this. I read in a mailing list that GIMP had similar problems with KDE and GNOME and that there are no defined standards for window managers. 

At least I could live with these minor issues.


Originally created by @franzlst  (Franz.Steinmetz@dlr.de) at 2016-09-08 11:46:33+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Same for me. Perhaps the problem vanishes in KDE 5 for OSleap and then we can enable full support for Gnome again. Furthermore before we go open source we should assure that we maximize the functionality for Ubuntu and Unity.


Originally created by @sebastian-brunner  (Sebastian.Brunner@dlr.de) at 2016-09-08 12:00:52+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
Okay then I will keep it like this for the moment and close the issue with the sidebars too.


Originally created by Benno  at 2016-09-08 12:01:37+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@Benno was not exactly what I thought we talked about but similar. :smile: 

For me the situation under KDE and sled11 is ok. And as long as I never maximize a sidebar but resize the situation in GNOME is ok, too. So we can leave it like it is!

The disappearing undocked widgets in KDE happens if any window is in front of any RAFCON window. The main window stays visible in this case. Hope that finishs all miss-understandings.

The feature it was talked about similar to pycharm can be explaned as follows. The floating mode for the sidebars in pycharm is undock mode in RAFCON. This floating windows can be hided by pressing a hide button in the up left corner of the widget or by pressing the destroy button in the up left corner of the frame. The un-hide function is triggered by e.g. pressing the remaining respective tab in the main-window. Finally, this should be understandable if you open pycharm and play with the sidebars and the floating mode. 

Finally, it is impressive that @Benno came to close to the same solution (focus strategies) for rafcon like in pycharm without knowing or using the floating mode feature in pycharm. :+1: 


Originally created by @Rbelder  at 2016-09-08 12:28:15+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
> The disappearing undocked widgets in KDE happens if any window is in front of any RAFCON window. The main window stays visible in this case. Hope that finishs all miss-understandings.

@Rbelder I understood the issue but the sidebars also disappear if the main window loses the focus and there is no other window in front of any RAFCON window. This could be fixed by just setting the undocked window type to gtk.gdk.NORMAL (just comment line 22 in undocked_window.py) but this will cause above stated error under GNOME.


Originally created by Benno  at 2016-09-08 12:50:52+00:00 (moved from RMC internal repository)

**Comment by franzlst** [comment]
@Benno Now, I got it, too. :smile: 


Originally created by @Rbelder  at 2016-09-08 12:53:20+00:00 (moved from RMC internal repository)
